### PR TITLE
Support quoted relations in collection

### PR DESCRIPTION
**Pull request by jesper-friis**
# Description
We use the collection as a local knowledge base. Since n3-represented literals contains embedded double-quotes it is essential that serialisation and de-serialisation of collections retain embedded double-quotes.

This PR ensures that embedded double-quotes are escaped in the internal representation of relations, which is used by the built-in json encoder/decoder.

## Type of change
- [x] Bug fix & code cleanup
- [x] New feature
- [ ] Documentation update
- [x] Test update

## Checklist for the reviewer
This checklist should be used as a help for the reviewer.

- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [ ] Is the code easy to read and understand?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?

### CODE REVIEW

File: src/dlite-type.c [hunk 1]
```diff
@@ -1198,23 +1204,50 @@ int dlite_type_scan(const char *src, int len, void *p, DLiteType dtype,
         return errx(dliteParseError, "relation should have 3 (optionally 4) elements");
       m = tokens->end - tokens->start;
       if (tokens->type == JSMN_ARRAY) {
+        size_t bufsize=0;
+        char *buf=NULL;
         if (!(t = jsmn_element(src, tokens, 0))) return -1;
-        rel->s = strndup(src + t->start, t->end - t->start);
+        if (strnput_unquote(&buf, &bufsize, 0, src + t->start,
```
Comment (quaat) [inline src/dlite-type.c:1210]: seems to be a lot of hard-to-read copy-pasted code here. Please create a function, or even a macro to simplify the reading.

File: src/utils/strutils.c [hunk 2]
```diff
@@ -268,14 +271,19 @@ int strnquote(char *dest, size_t size, const char *s, int n,
     if (size > i) dest[i] = '"';
     i++;
   }
-  while (s[j] && (n < 0 || (int)j < n)) {
-    if (s[j] == '"' && !(flags & strquoteNoEscape)) {
-      if (size > i) dest[i] = '\\';
+  if (s) {
```
Comment (quaat) [inline src/utils/strutils.c:274]: hard to follow the logic. Comment line-by-line what is happening here, or simplify the expression with standard functions.

File: src/utils/strutils.c [hunk 3]
```diff
@@ -342,6 +350,42 @@ int strnunquote(char *dest, size_t size, const char *s, int n,
 }
 
 
+/*
+  Like strnunquote(), but reallocates the destination and writes to
+  position `pos`.
+
+  On allocation error, -3 is returned.
+ */
+int strnput_unquote(char **destp, size_t *sizep, size_t pos, const char *s,
```
Comment (quaat) [inline src/utils/strutils.c:359]: Hard to follow logic. Please document what is the execution logic, or use a standard function to simplify the expression

**Review by quaat — CHANGES_REQUESTED** [review]

File: src/dlite-type.c (hunk 1 above)
Comment (jesper-friis) [reply src/dlite-type.c:1210]: Good point. Done

File: src/utils/strutils.c (hunk 2 above)
Comment (jesper-friis) [reply src/utils/strutils.c:274]: Added some comments to make it easier to follow.

File: src/utils/strutils.c (hunk 3 above)
Comment (jesper-friis) [reply src/utils/strutils.c:359]: Good suggestion. When adding explanations it became clear that the code both could be simplified and made more efficient by removing an unnecessary call to realloc().

**Review by quaat — APPROVED** [review]
